src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java

```diff
@@ -0,0 +1,429 @@
+      flushRequested = true;
+      synchronized (this) {
+        notifyAll();
+        for (long elapsedMs = time.milliseconds() - flushStartTimeMs;
+             !isTerminal() && elapsedMs < timeoutMs && bufferedRecords() > 0;
+             elapsedMs = time.milliseconds() - flushStartTimeMs) {
+          wait(timeoutMs - elapsedMs);
+        }
+        throwIfTerminal();
+        if (bufferedRecords() > 0) {
+          throw new ConnectException("Flush timeout expired with unflushed records: " + bufferedRecords());
+        }
+      }
+    } catch (InterruptedException e) {
+      throw new ConnectException(e);
+    } finally {
+      flushRequested = false;
+    }
+  }
+
+  private final class BulkTask implements Callable<BulkResponse> {
+
+    final long batchId = createdBatches.incrementAndGet();
+
+    final List<R> batch;
+
+    BulkTask(List<R> batch) {
+      this.batch = batch;
+    }
+
+    @Override
+    public BulkResponse call() throws Exception {
+      final BulkResponse rsp;
+      try {
+        rsp = execute();
+      } catch (Exception e) {
+        failAndStop(e);
+        throw e;
+      }
+      log.debug("Successfully executed batch {} of {} records", batchId, batch.size());
+      onBatchCompletion(batch.size());
+      return rsp;
+    }
+
+    private BulkResponse execute() throws Exception {
+      final B bulkReq;
+      try {
+        bulkReq = bulkClient.bulkRequest(batch);
+      } catch (Exception e) {
+        log.error("Failed to create bulk request from batch {} of {} records", batchId, batch.size(), e);
+        throw e;
+      }
+      for (int remainingRetries = maxRetries; true; remainingRetries--) {
+        boolean retriable = true;
+        try {
+          log.trace("Executing batch {} of {} records", batchId, batch.size());
+          final BulkResponse bulkRsp = bulkClient.execute(bulkReq);
+          if (bulkRsp.isSucceeded()) {
+            return bulkRsp;
+          }
+          retriable = bulkRsp.isRetriable();
+          throw new ConnectException("Bulk request failed: " + bulkRsp.getErrorInfo());
+        } catch (Exception e) {
+          if (retriable && remainingRetries > 0) {
+            log.warn("Failed to execute batch {} of {} records, retrying after {} ms", batchId, batch.size(), retryBackoffMs, e);
+            time.sleep(retryBackoffMs);
+          } else {
+            log.error("Failed to execute batch {} of {} records", batchId, batch.size(), e);
+            throw e;
+          }
+        }
+      }
+    }
+
+  }
+
+  private synchronized void onBatchCompletion(int batchSize) {
+    successfulBatches.incrementAndGet();
+    successfulRecords.addAndGet(batchSize);
+    inFlightRecords -= batchSize;
+    assert inFlightRecords >= 0;
+    notifyAll();
+  }
+
+  private void failAndStop(Throwable t) {
+    error.compareAndSet(null, toConnectException(t));
+    stop();
+  }
+
+  /**
+   * @return count of currently buffered records
+   */
+  public synchronized int unsentRecords() {
```

Comment:
A bunch of these accessors seem to be unused. Remove them until they are actually needed?